Add Tus docs (#29)
transloadit/uppy.io: docs/upload-strategies/tus.mdx

Keys are header names, values are header values.

```js
const headers = {
  authorization: `Bearer ${window.getCurrentUserToken()}`,
}
```

Header values can also be derived from file data by providing a function.
The function receives an [Uppy file][] and must return an object where the keys are header names, and values are header values.

```js
const headers = (file) => {
  return {
    authorization: `Bearer ${window.getCurrentUserToken()}`,
    expires: file.meta.expires,
  }
}
```

#### `chunkSize`

A number indicating the maximum size of a `PATCH` request body in bytes (`number`, default: `Infinity`).

:::caution
Do not set this value unless you are forced to.
The two valid reasons are described in the [tus-js-client docs](https://github.com/tus/tus-js-client/blob/master/docs/api.md#chunksize).
:::

#### `withCredentials`

Configure the requests to send Cookies using the [`xhr.withCredentials`](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/withCredentials) property (`boolean`, default: `false`).

The remote server must accept CORS and credentials.

#### `retryDelays`

> When uploading a chunk fails, automatically try again after the defined millisecond intervals (`array<number>`, default: `[0, 1000, 3000, 5000]`).

Reviewer: ```suggestion
When uploading a chunk fails, automatically try again after the defined millisecond intervals (`Array<number>`, default: `[0, 1000, 3000, 5000]`).
```
or
```suggestion
When uploading a chunk fails, automatically try again after the defined millisecond intervals (`number[]`, default: `[0, 1000, 3000, 5000]`).
```

#### `onShouldRetry: (err, retryAttempt, options, next)`

When an upload fails `onShouldRetry` is called with the error and the default retry logic as the last argument (`function`).

The default retry logic is an [exponential backoff](https://en.wikipedia.org/wiki/Exponential_backoff) algorithm triggered on HTTP 429 (Too Many Requests) errors.
Meaning if your server (or proxy) returns HTTP 429 because it’s being overloaded, @uppy/tus will find the ideal sweet spot to keep uploading without overloading.

If you want to extend this functionality, for instance to retry on unauthorized requests (to retrieve a new authentication token):

```js
import Uppy from '@uppy/core'
import Tus from '@uppy/tus'
new Uppy().use(Tus, {
  endpoint: '',
  async onBeforeRequest (req) {
    const token = await getAuthToken()
    req.setHeader('Authorization', `Bearer ${token}`)
  },
  onShouldRetry (err, retryAttempt, options, next) {
    if (err?.originalResponse?.getStatus() === 401) {
      return true
    }
    return next(err)
  },
  async onAfterResponse (req, res) {
    if (res.getStatus() === 401) {
      await refreshAuthToken()
    }
  },
})
```

#### `allowedMetaFields`

> Pass an array of field names to limit the metadata fields that will be added to uploads as [Tus Metadata](https://tus.io/protocols/resumable-upload.html#upload-metadata) (`array`, default: `null`).

Reviewer: ```suggestion
Pass an array of field names to limit the metadata fields that will be added to uploads as [Tus Metadata](https://tus.io/protocols/resumable-upload.html#upload-metadata) (`Array`, default: `null`).
```

Tus uses unique identifers for the file names to prevent naming collisions.
To still keep the meta data in place, Tus also uploads an extra `.info` file with the original file name and other meta data:

```json
{
  "ID":"00007a99d16d4eeb5a3e3c080b6f69da+JHZavdqPSK4VMtarg2yYcNiP8t_kDjN51lBYMJdEyr_wqEotVl8ZBRBSTnWKWenZBwHvbLNz5tQXYp2N7Vdol.04ysQAuw__suTJ4IsCljj0rjyWA6LvV4IwF5P2oom2",
  "Size":1679852,
  "SizeIsDeferred":false,
  "Offset":0,
  "MetaData":{
    "filename":"cat.jpg",
    "filetype":"image/jpeg"
  },
  "IsPartial":false,
  "IsFinal":false,
  "PartialUploads":null,
  "Storage":{
    "Bucket":"your-bucket",
    "Key":"some-key",
    "Type":"s3store"
  }
}
```

### How do I change files before sending them?

If you want to change the file names, you want to do that in [`onBeforeFileAdded`](/docs/uppy-core#onbeforefileaddedfile-files).

If you want to send extra headers with the request, use [`headers`](#headers) or [`onBeforeRequest`](#onbeforerequestreq-file).

### How do I change (or move) files after sending them?

If you want to preserve files names, extract meta data, or move files to a different place you generally can with hooks or events.

> It depends on the Tus server you use how it’s done exactly. [tusd][], for instance, exposes [hooks](https://github.com/tus/tusd/blob/master/docs/hooks.md)

Reviewer: ```suggestion
It depends on the Tus server you use how it’s done exactly. [`tusd`][], for instance, exposes [hooks](https://github.com/tus/tusd/blob/master/docs/hooks.md)
```

> and [tus-node-server](https://github.com/tus/tus-node-server) has [events](https://github.com/tus/tus-node-server#events).

Reviewer: ```suggestion
and [`tus-node-server`](https://github.com/tus/tus-node-server) has [events](https://github.com/tus/tus-node-server#events).
```

### Which server do you recommend?

[Transloadit](https://transloadit.com) runs [tusd][] in production, where it serves millions of requests globally.
Therefor we recommend tusd as battle-tested from our side, but other companies have had success with other [implementations][] so it depends on your needs.

### Are there hosted Tus servers?

> All [Transloadit plans](https://transloadit.com/pricing) come with a hosted [tusd][] server.

Reviewer: ```suggestion
All [Transloadit plans](https://transloadit.com/pricing) come with a hosted [`tusd`][] server.
```

You don’t have to do anything to leverage it, using [`@uppy/transloadit`](/docs/upload-strategies/transloadit) automatically
uses Tus under the hood.

### Why Tus instead of directly uploading to AWS S3?

<!--retext-simplify ignore first-and-foremost-->

First and foremost: reliable, resumable uploads. 
This means accidentally closing your tab or losing connection let’s you continue, for instance, your 10GB upload
instead of starting all over.

Tus is also efficient with lots of files (such as 8K) and large files.
Uploading to AWS S3 directly from the client also introduces quite a bit of overhead,
as more requests are needed for the flow to work.

> [tus-js-client]: https://github.com/tus/tus-js-client

Reviewer: ```suggestion
[`tus-js-client`]: https://github.com/tus/tus-js-client
```

[Uppy file]: /docs/uppy-core#working-with-uppy-files

[tus]: https://tus.io/

> [tusd]: https://github.com/tus/tusd

Reviewer: ```suggestion
[`tusd`]: https://github.com/tus/tusd
```

---
sidebar_position: 2
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import UppyCdnExample from '/src/components/UppyCdnExample';

# Tus

The `@uppy/tus` plugin brings resumable file uploading with [Tus](http://tus.io) to Uppy by wrapping the [tus-js-client][].

## When should I use it?

:::tip
Not sure which upload strategy is best for you? Read “[Choosing the upload strategy you need](/docs/guides/choosing-upload-strategy)”.
:::

[Tus][tus] is an open protocol for resumable uploads built on HTTP.

> This means accidentally closing your tab or losing connection let’s you continue, for instance, your 10GB upload
> instead of starting all over.

Reviewer 2: ```suggestion
This means that accidentally closing your tab or losing connection lets you continue, for instance, your 10GB upload instead of starting all over.
```

> It supports any language, any platform, and any network. You can checkout the [implementations](https://tus.io/implementations.html) to
> find the implementation in your favorite language. You can store files on your self-hosted Tus server, but also use service integrations (such as S3) to store files
> externally.

Reviewer 2: ```suggestion
Tus consists of a client and a server. Multiple client and server [implementations](https://tus.io/implementations.html) exist in many languages. Uppy uses `tus-js-client` internally and you can set up any server you like to use with Uppy. You can store files on your self-hosted Tus server itself, but more common is to use service integrations (such as S3) to store files externally.

:::tip
If you don't want to host your own Tus server, you can take advantage of Transloadit's hosted services - read “[Choosing the upload strategy you need](/docs/guides/choosing-upload-strategy)”.
:::
```

Author: I'm not sure we need to reiterate this here. At the end of the day, this is still an OSS product and the website as well. The guide is already linked 10 lines up and the fact that TL has hosted servers is also mentioned in FAQ. What do you think??

Author: I would be more open to this change if we don't use `:::tip` and we don't link to the same thing as the previous tip does. Perhaps this can be one extra line, with a link to the FAQ below?

Reviewer 2: good idea! i updated the comment

Author: I think I still prefer the previous text and would only add this new sentence:

```suggestion
It supports any language, any platform, and any network. You can checkout the [implementations](https://tus.io/implementations.html) to
find the implementation in your favorite language. You can store files on your self-hosted Tus server, but also use service integrations (such as S3) to store files
externally. If you don't want to host your own Tus server, you can take advantage of [Transloadit's hosted services](#are-there-hosted-tus-servers). 
```

Reviewer 2: I thought it would be worth mentioning that Tus is a server and a client, because newcomers might be confused as to why they would choose tus in any language when clearly Uppy is a Javascript based library. Whether Tus client supports many different languages/platforms is not so interesting for someone who just wants to get Uppy up and running.

Author: > newcomers might be confused as to why they would choose tus in any language when clearly Uppy is a Javascript based library

People first need to setup a Tus server before they can do anything. Just because Uppy is JS, doesn't mean that their backend is, they may have an existing .NET backend. I think it's important to reference the implementations to choose from.

Reviewer 2: yes, that's why i want to mention that tus is a client and a server, so that they understand that they must choose a tus server, not a client

Reviewer 3: I agree with @mifi here. The docs should mention that you need a tus server to use the protocol and plugin for uploading. For of the time, this is clear. But from time to time users still get confused.

Author: I incorporated your suggestions but made it a bit more succinct. This is the "Why should you use it section" and some details (like tus-js-client) are already provided in the section above.

```suggestion
Tus supports any language, any platform, and any network. It requires a client and server integration to work. You can checkout the client and server [implementations](https://tus.io/implementations.html) to
find the server in your preferred language. You can store files on the Tus server itself, but you can also use service integrations (such as S3) to store files externally. If you don't want to host your own server, see “[Are there hosted Tus servers?](#are-there-hosted-tus-servers)”.
```

If you want reliable, resumable uploads: use `@uppy/tus` to connect to your Tus server in a few lines of code.

## Install

<Tabs>
  <TabItem value="npm" label="NPM" default>

  ```shell
  npm install @uppy/tus
  ```

  </TabItem>

  <TabItem value="yarn" label="Yarn">

  ```shell
  yarn add @uppy/tus
  ```

  </TabItem>

  <TabItem value="cdn" label="CDN">
    <UppyCdnExample>
      {`
        import { Uppy, Tus } from "{{UPPY_JS_URL}}"
        new Uppy().use(Tus, { endpoint: 'https://tusd.tusdemo.net/files' })
      `}
    </UppyCdnExample>
  </TabItem>
</Tabs>

## Use

A quick overview of the complete API.

```js {10} showLineNumbers
import Uppy from '@uppy/core'
import Dashboard from '@uppy/dashboard'
import Tus from '@uppy/tus'

import '@uppy/core/dist/style.min.css'
import '@uppy/dashboard/dist/style.min.css'

const uppy = new Uppy
  .use(Dashboard, { inline: true, target: 'body' })
  .use(Tus, { endpoint: 'https://tusd.tusdemo.net/files/' })
```

## API

### Options

:::info

> All options are passed to `tus-js-client` and we document the ones here that we added or changed.

Reviewer 3: If you say that you only mention options differing from tus-js-client, then why does it mention endpoint,, overridePatchMethod, and retryDelays?

Author: Good point! I should reword that we also document required options, like endpoint. The `retryDelays` default is different than the tus-js-client default. `overridePatchMethod` can indeed be removed.
